test/unit/math/mix/meta/eigen_utils.hpp

```diff
@@ -0,0 +1,298 @@
+#ifndef TEST_UNIT_MATH_MIX_EIGEN_UTILS_HPP
+#define TEST_UNIT_MATH_MIX_EIGEN_UTILS_HPP
+#include <stan/math/mix.hpp>
+#include <Eigen/Sparse>
+#include <gtest/gtest.h>
+
+namespace stan {
+namespace math {
+namespace test {
+
+/**
+ * Test type checking for sparse matrices
+ * @tparam eigen_base_v logical for whether `Checker` should pass for
+ *   `EigenBase` types.
+ * @tparam eigen_sparse_compressed_v logical for whether `Checker` should pass
+ * for `SparseCompressedBase` types.
+ * @tparam eigen_sparse_matrix_v logical for whether `Checker` should pass for
+ *   `SparseMatrixBase` types.
+ * @tparam eigen_sparse_map_v logical for whether `Checker` should pass for
+ *   `SparseMapBase` types.
+ * @tparam Scalar The scalar type of the Eigen object to check.
+ * @tparam Checker A struct taking in an expression template type and returning
+ *  a true if check is satisfied and false otherwise.
+ */
+template <bool eigen_base_v, bool eigen_sparse_compressed_v,
+          bool eigen_sparse_matrix_v, bool eigen_sparse_map_v, typename Scalar,
+          template <class...> class Checker>
+void test_eigen_sparse_matrix() {
+  using Eigen::EigenBase;
+  using Eigen::SparseCompressedBase;
+  using Eigen::SparseMapBase;
+  using Eigen::SparseMatrix;
+  using Eigen::SparseMatrixBase;
+  using SparseMatrixXd = SparseMatrix<Scalar>;
+  SparseMatrixXd C(10, 10);
+  SparseMatrixXd D(10, 10);
+  // scalars
+  EXPECT_FALSE((Checker<bool>::value));
+  EXPECT_FALSE((Checker<double>::value));
+  EXPECT_FALSE((Checker<int>::value));
+
+  // Sparse
+  EXPECT_TRUE((eigen_base_v == Checker<EigenBase<decltype(C * D)>>::value));
+  EXPECT_TRUE((eigen_sparse_compressed_v
+               == Checker<SparseCompressedBase<decltype(C * D)>>::value));
+  EXPECT_TRUE((eigen_sparse_matrix_v
+               == Checker<SparseMatrixBase<decltype(C * D)>>::value));
```

Comment:
Why do we need checking with base classes? They never explicitely instantinated in our code (only in eigen functions taking them as arguments). We are only using derived types - only those need to be checked.

I think we need to check with a plain type (for example `declytpe(C)`) and some expression (for example `decltype(C+D)`). Map is fine too (actually, does map for sparse matrix even make sense?).

Reply:
> Why do we need checking with base classes? They never explicitly instantiated in our code (only in eigen functions taking them as arguments). We are only using derived types - only those need to be checked.

What about something like `(A * B).matrix()`? That get's wrapped in a MatrixBase

https://godbolt.org/z/TFXJ5Q

> I think we need to check with a plain type (for example declytpe(C)) and some expression (for example decltype(C+D)). Map is fine too (actually, does map for sparse matrix even make sense?).

`decltype(C)` would be the same type as `C` no? Which would be covered by the check on `test_mat`

https://godbolt.org/z/jj6Ywc

> Map is fine too (actually, does map for sparse matrix even make sense?).

I think so? SparseMapBase is a thing. I could add a test for just plain `Map`

Reply:
> What about something like (A * B).matrix()?

Nice catch. I did not know about this one. I still think we can't get `EigenBase`. Or can we?

> decltype(C) would be the same type as C no? Which would be covered by the check on test_mat

Right.

> I think so? SparseMapBase is a thing. I could add a test for just plain Map

I will leave that up to you. I am not very familiar with sparse matrices.